ast_tools/passes/ssa.py

```diff
@@ -283,7 +284,8 @@ def _prove_names_defined(
     if isinstance(node, ast.Name):
         if isinstance(node.ctx, ast.Store):
             names.add(node.id)
-        elif node.id not in names and node.id not in env:
+        elif node.id not in names and node.id not in env and \
+                node.id not in dir(builtins):
```

Comment:
I think better would be to check:
```
if '__builtins__' in env:
   builtins_d = env['__builtins__'].__dict__
else:
   builtins_d = builtins.__dict__
...
elif ... node.id not in buildins_d: ...
```
This would match the behavior of `exec` allowing the developer to choose which builtins are available by setting `'__builtins__'`. https://docs.python.org/3/library/functions.html#exec

Reply:
I changed it to just use `env["__builtins__"].__dict__` always because it seems like that should always be defined the environment (whether default or user defined) if we're starting with a call to `locals` and/or `globals`, e.g.
``
❯ python
Python 3.7.1 (default, Dec 14 2018, 13:28:58)
[Clang 4.0.1 (tags/RELEASE_401/final)] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> globals()
{'__name__': '__main__', '__doc__': None, '__package__': None, '__loader__': <class '_frozen_importlib.BuiltinImporter'>, '__spec__': None, '__annotations__': {}, '__builtins__': <module 'builtins' (built-in)>}
>>> locals()
{'__name__': '__main__', '__doc__': None, '__package__': None, '__loader__': <class '_frozen_importlib.BuiltinImporter'>, '__spec__': None, '__annotations__': {}, '__builtins__': <module 'builtins' (built-in)>}
```